Feat: comp_multicommand

Trico-Everfire opened the pull request:
the comp_multicommand postcomp entities are a simplifcation of calling multiple console commands without needing to set up a `logic_relay` + `point_client/servercommand` entity combo (It does this internally) There are 2 types of multi command entities for each of the 2 types of commands (So 4 entries total).
`comp_multi_server/client_command`
and
`comp_auto_multi_server/client_command`

`comp_multi_server/client_command` works similar to a trigger (because it compiles to a trigger) that when triggered will execute the commands set in the entity (the user can specify up to 16 individual commands if they wish).

`comp_auto_multi_server/client_command` Is not triggerable. Rather is executed by itself on spawn and will also delete itself afterwards to save on entity count but otherwise acts the same as the normal `comp_multi_server/client_command`

Trico-Everfire committed:
add: comp multi command

vrad-exe commented:
There is actually a third command entity, `point_broadcastclientcommand`, which sends the command to all players in multiplayer instead of just the activator. Instead of adding 2 more extra postcompiler entities for that it would probably make more sense to just combine them all together with a keyvalue to choose the command type.

TeamSpen210 commented:
Similarly I also just recently added `comp_movie_fitter`, which has a mode that swaps between on-trigger and on-spawn.

Trico-Everfire commented:
I see now. I'll see what I can do. This all makes sense.

Trico-Everfire committed:
Add/Change: additions/changes based on feedback. Also icon added.

Trico-Everfire commented:
Provided icons for the entity (it's now 1 as suggested). Icon provided by Ashanderite on Discord (provided specifically for this entity.)

TeamSpen210 committed:
Formatting

TeamSpen210 committed:
Remove redundant check

If there aren't any multi commands, the loop just does nothing

TeamSpen210 committed:
Simplify gathering commands

Iterate all keyvalues, this way users can turn off SmartEdit to include any number of keys

TeamSpen210 committed:
Fix command locating logic

Look at all entities of the correct type, see if any have a unique name.

TeamSpen210 committed:
Tweak FGD

TeamSpen210 commented:
Fixed/tweaked a couple things, see the commits for info. Only thing I'm not sure about is the `;` joining - there's the danger of hitting the I/O parameter text limit (I think 255 characters). Could track the length, split if it's going to exceed, or just always produce one output per keyvalue.

Your existing code also did ensure ordering is preserved, but it didn't allow non-consecutive keys or ones beyond the predefined ones. Maybe it'd be worth extracting the number, sorting by that?

Trico-Everfire commented:
I didn't know IO had limits to command length (I hate this engine sometimes). Splitting by command might be for the best. As for order. I do think it matters? Sometimes some commands rely on others to exist. Like how every cheat command relies on `sv_cheats`. I don't know if other commands are like this.

Trico-Everfire commented:
I have a piece of criticism. I purposfully tried to avoid this:
```py
    for command in vmf.by_class[classname]:
        name = command['targetname']
        if name and len(vmf.by_target[name]) == 1:
            # It's named uniquely, use it.
            return command
```
We are grabbing an entity of unknown origin. Cool. We don't know where it came from or who owns it. If someone, for who knows what reason, deletes this entity. comp_multicommand breaks. I know it's unlikely someone were to delete a point_<type>command entity. But if they ever feel the need to, maybe to keep entity limits down after the entity served it's purpose. We're now stuck. comp_multicommand doesn't own the entity. Which can lead to problems.

TeamSpen210 commented:
Definitely is a concern, though your existing code wouldn't have worked. It seemed to do the same, but without checking? In that case, probably just make a dict at the start of the context function, store a reference to the created command entities.

Trico-Everfire commented:
Mine did account for that. It checked if the entity named the unique name specific to this entity that if it existed, it used it, otherwise created it:
```py
    command = ctx.vmf.by_target[entity_name]

    if(len(command) < 1): # If it doesn't exist, create one.
        command = ctx.vmf.create_ent(entity_name)
        command["targetname"] = f'cmp_multi_{entity_name}'
```
I do realize I made a mistake with it.
This should be it:
```py
    command = ctx.vmf.by_target[f'cmp_multi_{entity_name}']

    if(len(command) < 1): # If it doesn't exist, create one.
        command = ctx.vmf.create_ent(entity_name)
        command["targetname"] = f'cmp_multi_{entity_name}'
```

TeamSpen210 commented:
Ahh, that makes sense.

TeamSpen210 committed:
Only use ents created by comp_multi_command

TeamSpen210 committed:
Preserve command order, account for max output limit

Trico-Everfire commented:
Tested. Seems good to go.
